### DISCUSSION TITLE
Force GCC when building bootloader.

### DISCUSSION THREAD

**Discussion by Blank-c** [pyinstaller/pyinstaller#7783] [PyInstaller] [answered]
I have both MSVC and GCC. MSVC has a lot of false positives in my case, even for simple C++ programs. But GCC don't give me those problems. So my question is, how can I force the wafscript to compile the bootloader using GCC without modifying Paths?

Does gcc compiles it with `-static` flag?

If I were to modify the MAGIC base, do I have to keep its length or some of bytes' pattern same? Or it can be anything like `b'COOKIE\x1\x2\x3'`?

**Comment by bwoodsend** [comment]
Set the `CC` environment variable to `gcc` to make it use gcc.

**Reply by rokm** [reply, marked as answer]
If there is MSVC available, I think you need to force gcc via `--gcc`  flag (`python waf all --gcc`), otherwise we end up resolving and using MSVC first.

> How does it search for GCC?

Using whatever search mechanism(s) `waflib` is employing. One of those is looking at the `CC` environment variable. Don't know about others (if any), though.

> What if gcc is not added to path, would it fail?

Probably.

But you should be able to set full path to the gcc executable in the `CC` environment variable.

> Does `gcc` compiles it with -static flag?

I don't think so.

> If I were to modify the MAGIC base, do I have to keep its length or some of bytes' pattern same?

Generally, there should be no restrictions on the MAGIC pattern. With changed length you might run into issues with structure's alignment on some platforms, though.

**Reply by Blank-c** [reply]
So since the original size of bytes is 8 bytes
```c
const unsigned char MAGIC_BASE[8] = {
    'M', 'E', 'I', 000,
    013, 012, 013, 016
};
``` 
So, I can put in their fully random 8 bytes?

But in [this](https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/archive/writers.py#L120) line, it uses a different magic pattern
```python
_COOKIE_MAGIC_PATTERN = b'MEI\014\013\012\013\016'
```
Why does this uses a different cookie than the magic base than in bootloader? How can I modify it to be used with the custom 8 bytes I provide?

**Reply by rokm** [reply]
Because magic "base" is literally a base and not final pattern used in cookie search:

https://github.com/pyinstaller/pyinstaller/blob/79f62ef29822169ae00cd4271390d0e3175476ad/bootloader/src/pyi_archive.c#L391-L396

 We cannot store the pattern directly in the bootloader, because then it would appear in the executable even when the actual PKG/CArchive is missing.

**Reply by Blank-c** [reply]
Ah got it, thanks

**Comment by Blank-c** [comment]
How does it search for GCC? What if gcc is not added to path, would it fail?

And please answer my other questions above if you can. I would appreciate that.
